### ISSUE TITLE
Restructure content

### ISSUE THREAD

**Issue by jhnns** [webpack/docs#64]
We should make the documentation more beginner-friendly. This issue is dedicated to a discussion about how we could restructure the content. Proposals welcome :)

However, we should still address the needs of advanced users (webpack/webpack#1627)

**Comment by leandroruel** [comment]
i agree, at current documentation, some people make jokes with memes at comments area, because the documentation don't have a complete "getting started"

**Comment by Shinobi881** [comment]
I'd be great to have better docs on use with express, HMR...
I'll try to come up with more specific examples/suggestions...

In terms of addressing needs more specific to user experience, maybe just create *beginner, intermediate, advanced* sections with brief descriptions of what a user should understand about the following section. 

Thanks for opening the conversation.

**Comment by gregvenech** [comment]
Over the past weekend I sketched out and put together a mockup for what IMHO would be a solid redesign for the Webpack documentation. I tried to use many of the same design elements (fonts, colors, cube icon) but cleaned up the layout and restructured the content. 

I included a PNG of the mockup below, and I'm obviously willing to share the PSD if anyone's interested. Using the same theme, I'd also be willing to put together a design for the splash page (and maybe the analyze tool) as well. I figured I should share this mockup before doing anything else. 

Webpack has completely changed my workflow and I feel this is the least I can do to help. Lemme know what you think, thanks!

**NOTE:** _I included real content from the docs but did leave a fair amount out. This is partly intentional because I think some of it is a bit wordy, but also a factor of my limited knowledge in certain areas of the tool. Just a rough draft..._

![documentation](https://cloud.githubusercontent.com/assets/8988697/15765524/c1d808f8-2903-11e6-8834-f0afbdd32bed.png)

**Comment by jhnns** [comment]
Awesome! That looks great 👍 

Since you've obviously made some thoughts about the structure of the content, could you share your intentions with us? The design is important, but I feel that rewriting and restructuring the content is even more important. We need to make things easier for beginners (by leaving out unimportant stuff), but we should not remove any content for experts.

I really like your general structure. It looks pretty similar to my plans about the content:

- **Guides**: Contains Introduction, Loaders, Code Splitting, Optimization, HMR etc. Generally the wordy stuff which is intended as a tutorial. Maybe we could split it into lessons which makes it easier to get a bigger picture of webpack gradually. These guides are not just for beginners, but they are intended to introduce the concept behind webpack.
- **Reference**: Provides concise descriptions of the `webpack.config.js`, the Node API and the CLI, the webpack-dev-middleware and the webpack-dev-server
- **Contribute**: Provides internal documentation to ease contribution
- **Analyze**: This tool should definitely be part of the official website.

This is my feedback regarding the design:

- I think the typography, especially the body text, is a little bit too small. You could try out a bigger font-size, shorter line-lengths and increase the leading. Maybe you could use a [modular scale](http://zellwk.com/blog/modular-scale/)?
- I haven't talked with @sokra about this, but I'd change the typography and polish the logo a little bit. I'm missing the remarkable bits of the design, like a special font face or a special color.
- Anyway, it's good to get things going. Your suggestions (and contributions) are welcome! :)

Things we needs to sort out:

@bebraw is [currently working](http://survivejs.com/blog/survivejs-webpack-120/) on [a book](http://survivejs.com/webpack/) about webpack. He is really doing a great job and (afaik) many parts of the book will be publicly available. I think a book is great and it's good to have a deep introduction, but I'd really like to have some guides *on* the website. Maybe rather short guides on the site while providing a link for the deep-dive? Could we copy stuff from your book or is it better to it write from scratch? What do you think @bebraw?

**Comment by bebraw** [comment]
The design looks good to me. I agree on the font size, maybe bump that a notch.

> @bebraw is currently working on a book about webpack. He is really doing a great job and (afaik) many parts of the book will be publicly available. I think a book is great and it's good to have a deep introduction, but I'd really like to have some guides on the website. Maybe rather short guides on the site while providing a link for the deep-dive? Could we copy stuff from your book or is it better to it write from scratch? What do you think @bebraw?

That's a good question.

The book content will likely evolve somewhat. Copying might give you starting points but eventually the content will probably diverge (too painful to keep in sync).

No matter what the structure will end up being, someone has to develop all the content. It would be  worthwhile to try to solve a small part of it programmatically. If we had a well annotated schema of configuration, you could generate up to date documentation based on that. So when the implementation changes, so do the docs. You lose in flexibility but get maintainability in return.

A schema would help in other ways too as it would enable people to develop tooling on top of it. The current schema is implicit (literally code). In webpack-validate we had to implement a custom schema of our own to validate it. So if we could push that closer to the core, that would be great. This would be one way to alter the definition of the problem.

When it comes to loaders and plugins, something similar is possible there too if you consider npm as your registry and generate lists for these based on meta information (keywords, description, readme). Basic listings are simple. Better UX requires higher level of standardization. But again, doing this would push the problem to the right place and reduce the pain of documenting.

I agree the main documentation should contain different level docs. Each user comes with needs of own after all and those needs may change as you use webpack. Considering this the structure you proposed (guides, reference, contribute, analyze) seems sane to me. For the sake of consistency I might name the categories differently (use, refer, contribute, analyze?) but I like the idea.

When it comes to guides, it could be enough to have enough documentation to get started and then point to other directions for more. Reference could be handled as I discussed above. Contribution related docs are the biggest weak point at the moment. That's also the hardest to get together. I agree on the analyze tool idea. It should be an integral part of the site.

**Comment by gregvenech** [comment]
Hey thanks for getting back to me so quickly... I'm glad you guys like it! In terms of restructuring these were my thoughts (referring to sections in the current docs):

**Guides**
- *Getting Started*, I think we should try to stick to top-level overviews here or maybe make a "concepts" section.
- *Tutorials and Examples*
- *Guides*, with the exception of *How to write a Loader/Plugin* which I think should be moved to "Contribute" (or "Development" in the mockup). 
- From *Lists* - The *List of Tutorials* and *List of Hints*
- *Webpack With* can probably go in this section as well

**Documentation**
- Everything from *API*, basically using those as the top level nodes in the side nav.
- From *Lists* - The *List of Loaders* and *List of Plugins* I think should go under the "Plugins"/"Loaders" top level nodes in the side nav.

**Contribute** (or Development)
- *Development* - *Changelog* and *Roadmap*, I think the *Ideas* section can maybe be removed and instead just point people to the issues section on Github. (I also agree with @shama on #63)
- *How to write a Loader/Plugin*, though I guess this could also go under Plugins or Loaders in the documentation

**Splash page and Analyze Tool**
- I definitely think the [splash page](http://webpack.github.io/) should be merged into this main site basically as the home page. 
- Because of it's complexity (I assume although I haven't poked around in that repo much), keeping the Analyze tool in separate repo probably makes sense. However I think branding it, maybe using same banner and footer, and prominently linking to it would make a huge difference in how many people try it out.

**Design**
> I think the typography, especially the body text, is a little bit too small. You could try out a bigger font-size, shorter line-lengths and increase the leading. Maybe you could use a modular scale?

I can definitely tweak the font size a bit, I hadn't heard of the Modular Scale plugin but it sounds like a great idea.

> I haven't talked with @sokra about this, but I'd change the typography and polish the logo a little bit. I'm missing the remarkable bits of the design, like a special font face or a special color.

I hear you. I used the same fonts as the current site both from Google Fonts. I don't think the Averia Sans Libre looks terrible as a logotype but there is room for improvement, maybe we can look on TypeKit or somewhere else. *On the other hand*, I really like the cube and animation... I think we could add a hover behavior (where it spins and maybe the inner cube changes color or something) to make it even better.

**Comment by sokra** [comment]
> I haven't talked with @sokra about this

Feel free to change everything...

---

For the docs it's probably a good idea to start from stretch and just ignore the old content.

With @bebraw s book we are in the good situation in which we can just omit detailed tutorials and instead point directly to the book. I don't think copying is fair (or legal?). @bebraw did a great job and as it's also available for free we can just point to it (and recommend buying to be fair :smile: ).

We could still offer some basic tutorials to getting started.

The guides section could include some in-depth guides of some aspects. Similar to this one but with a bit more explainations: https://gist.github.com/sokra/ff1b0290282bfa2c037bdb6dcca1a7aa Here we could point to some examples too.

And we need to include a API documentation, which sometimes point to a relevant guide that explains the option in more detail. (Maybe partially autogenerated from a schema)

**Comment by bebraw** [comment]
> With @bebraw s book we are in the good situation in which we can just omit detailed tutorials and instead point directly to the book. I don't think copying is fair (or legal?). @bebraw did a great job and as it's also available for free we can just point to it (and recommend buying to be fair 😄 ).

The content is available under CC BY-NC-ND license, but I think I can flex on that especially when it comes to graphics. I need to get a series of illustrations done and sharing those could be a nice touch. There's certain level of overlap after all.

> We could still offer some basic tutorials to getting started.

Yeah. I'm thinking having a good 5-10 min tutorial would be good for those starting out. Just go through the basic concepts and mindset. The current site has something already iirc, but we can likely do even better.

**Comment by also** [comment]
For the "Development" or "Contribute" section, how do people feel about including some JSDoc-style generated documentation? While I think that these generated docs are not great when they're the only API documentation, they are a good place to provide a short description and link to something more comprehensive.

I've found myself referring to https://github.com/webpack/docs/wiki/internal-webpack-plugins while debugging through the source, so I converted that to simple JSDoc comments: https://github.com/webpack/webpack/compare/master...also:jsdoc ([published with the default documentation.js theme](https://also.github.io/webpack/internal-plugins/)).

If this seems like it could be useful, I'm happy to clean up those JSDoc comments, PR them, and work on adding more and integrating them into the site.

Otherwise, are there better ways to document this sort of thing? Markdown seems a bit limited.

**Comment by jhnns** [comment]
> While I think that these generated docs are not great when they're the only API documentation, they are a good place to provide a short description and link to something more comprehensive.

Dito. I don't like generated docs because usually they just describe the API, not the concepts. But it's a good idea to provide these for potential contributors. Since these docs are about internal stuff, they may change frequently and it's important to have an up-to-date overview of the internals.

> We could still offer some basic tutorials to getting started.

I think that's a good idea. The site could offer compact introductions to the concepts and link to a chapter of the book.

**Comment by gregvenech** [comment]
@jhnns got back to designing a bit more this weekend... should have more shots of mockups tomorrow night or Monday (including Splash/Home page). I'm thinking maybe I should start a new issue where we can discuss more design specific stuff and leave this thread for restructuring talk?

**Comment by SpaceK33z** [comment]
Closing since this is being done in https://github.com/webpack/webpack.js.org 💯
